server/tracedb/elasticsearchdb.go

```diff
@@ -0,0 +1,326 @@
+				Error:                err,
+			},
+		}
+	}
+
+	return connectionTestResult
+}
+
+func (db elasticsearchDB) Ready() bool {
+	return db.client != nil
+}
+
+func (db elasticsearchDB) GetTraceByID(ctx context.Context, traceID string) (model.Trace, error) {
+	if !db.Ready() {
+		return model.Trace{}, fmt.Errorf("ElasticSearch dataStore not ready")
+	}
+	content := strings.NewReader(fmt.Sprintf(`{
+		"query": { "match": { "trace.id": "%s" } }
+	}`, traceID))
+
+	searchRequest := esapi.SearchRequest{
+		Index:  []string{db.config.Index},
+		Body:   content,
+		Pretty: true,
+	}
+
+	response, err := searchRequest.Do(ctx, db.client)
+	if err != nil {
+		return model.Trace{}, fmt.Errorf("could not execute search request: %w", err)
+	}
+	defer response.Body.Close()
+
+	responseBody, err := ioutil.ReadAll(response.Body)
+	if err != nil {
+		return model.Trace{}, fmt.Errorf("could not read response body")
+	}
+
+	var searchResponse searchResponse
+	err = json.Unmarshal(responseBody, &searchResponse)
+	if err != nil {
+		return model.Trace{}, fmt.Errorf("could not unmarshal search response into struct: %w", err)
+	}
+
+	if len(searchResponse.Hits.Results) == 0 {
+		return model.Trace{}, ErrTraceNotFound
+	}
+
+	return convertElasticSearchFormatIntoTrace(traceID, searchResponse), nil
+}
+
+func newElasticSearchDB(cfg *config.ElasticSearchDataStoreConfig) (TraceDB, error) {
+	var caCert []byte
+	if cfg.Certificate != "" {
+		caCert = []byte(cfg.Certificate)
+	}
+
+	client, err := elasticsearch.NewClient(elasticsearch.Config{
+		Addresses: cfg.Addresses,
+		Username:  cfg.Username,
+		Password:  cfg.Password,
+		CACert:    caCert,
+	})
+	if err != nil {
+		return nil, fmt.Errorf("could not create elasticsearch client: %w", err)
+	}
+
+	return &elasticsearchDB{
+		config: cfg,
+		client: client,
+	}, nil
+}
+
+func getClusterInfo(client *elasticsearch.Client) (string, error) {
+	var r map[string]interface{}
+	res, err := client.Info()
+	if err != nil {
+		return "", fmt.Errorf("error getting cluster info response: %s", err)
+	}
+	defer res.Body.Close()
+
+	// Check response status
+	if res.IsError() {
+		return "", fmt.Errorf("error getting cluster info response status: %s", res.String())
+	}
+	// Deserialize the response into a map
+	if err := json.NewDecoder(res.Body).Decode(&r); err != nil {
+		return "", fmt.Errorf("error parsing cluster info response: %s", err)
+	}
+
+	// Return client version number
+	info := fmt.Sprintf("Server: %s", r["version"].(map[string]interface{})["number"])
+	return info, nil
+}
+
+func convertElasticSearchFormatIntoTrace(traceID string, searchResponse searchResponse) model.Trace {
+	spans := make([]model.Span, 0)
+	for _, result := range searchResponse.Hits.Results {
+		span := convertElasticSearchSpanIntoSpan(result.Source)
+		spans = append(spans, span)
+	}
+
+	return model.NewTrace(traceID, spans)
+}
+
+func convertElasticSearchSpanIntoSpan(input map[string]interface{}) model.Span {
```

Comment:
Question (not important, out of curiosity): is there a reason on why we are parsing a span here using `flatten` and getting thought a map?

Other possibility it was create a temporary type for the Elastic span and use `json.Unmarshal` to get the values, in a code similar to this:

```golang
type elasticDuration struct {
    Us float64 `json:"us"`
}

type elasticCommonInfo struct {
    Id string `json:"id"`
    Name string `json:"name"`
    Duration `json:"duration"`
}

type elasticSearchSpan struct {
    Transaction elasticCommonInfo `json:"transaction"`
    Span elasticCommonInfo `json:"span"`
    Timespan string `json:"@transaction"`
}

// to do that we need to think how to bring input as an array of bytes (maybe using json.RawMessage?)
func convertElasticSearchSpanIntoSpan(input []byte) model.Span {
    var esSpan elasticSearchSpan 
    err := json.Unmarshal(input, &esSpan)

    // after handling error we could retrieve this values like this
    transactionId := esSpan.Transaction.Id
    spanId := esSpan.Span.Id

    transactionName := esSpan.Transaction.Name
    spanName := esSpan.Span.Name

    transactionDuration := esSpan.Transaction.Duration.Us
    spanDuration := esSpan.Span.Duration.Us

    timespan = esSpan.Timespan
}

```

Reply:
Talked with @danielbdias about this. I think we can go with the current approach (flatten + map). The thing about creating a struct is that we don’t know all the fields of the span, also maybe it is not good try to parse the fields that we don't know, because they can change through elastic versions.